frb_dart/lib/src/ffi/io.dart

```diff
@@ -51,26 +37,134 @@ class DartApiDl {
+            (i) => value.as_array.values.elementAt(i).value.ref.intoDart());
+
+      case 7: // DartTypedData
+        return _typedDataIntoDart(
+          value.as_typed_data.ty,
+          value.as_typed_data.values,
+          value.as_typed_data.length,
+          copy: true,
+        );
+
+      case 8: // DartExternalTypedData
+        final externalTypedData = _typedDataIntoDart(
+          value.as_external_typed_data.ty,
+          value.as_external_typed_data.data,
+          value.as_external_typed_data.length,
+          copy: false,
+        );
+        _externalTypedDataFinalizer.attach(
+            externalTypedData, value.as_external_typed_data);
+        return externalTypedData;
+
+      case 9: // DartSendPort
+      case 10: // DartCapability
+      case 11: // DartNativePointer
+      case 12: // DartUnsupported
+      case 13: // DartNumberOfTypes
+      default:
+        throw "Can't read invalid data type $ty";
     }
-    return Uint8List.fromList(ptr.asTypedList(len));
   }
 
-  bool get isSuccess => success > 0;
+  static dynamic _typedDataIntoDart(
+    int ty,
+    ffi.Pointer<ffi.Void> typedValues,
+    int nValues, {
+    required bool copy,
+  }) {
+    switch (ty) {
+      case 0: // ByteData
+        final view = typedValues.cast<ffi.Uint8>().asTypedList(nValues);
+        final bytes = copy ? view : Uint8List.fromList(view);
+        return ByteData.view(bytes.buffer);
+      case 1: // Int8
+        final view = typedValues.cast<ffi.Int8>().asTypedList(nValues);
+        return copy ? Int8List.fromList(view) : view;
+      case 2: // Uint8
+        final view = typedValues.cast<ffi.Uint8>().asTypedList(nValues);
+        return copy ? Uint8List.fromList(view) : view;
+      case 4: // Int16
+        final view = typedValues.cast<ffi.Int16>().asTypedList(nValues);
+        return copy ? Int16List.fromList(view) : view;
+      case 5: // Uint16
+        final view = typedValues.cast<ffi.Uint16>().asTypedList(nValues);
+        return copy ? Uint16List.fromList(view) : view;
+      case 6: // Int32
+        final view = typedValues.cast<ffi.Int32>().asTypedList(nValues);
+        return copy ? Int32List.fromList(view) : view;
+      case 7: // Uint32
+        final view = typedValues.cast<ffi.Uint32>().asTypedList(nValues);
+        return copy ? Uint32List.fromList(view) : view;
+      case 8: // Int64
+        final view = typedValues.cast<ffi.Int64>().asTypedList(nValues);
+        return copy ? Int64List.fromList(view) : view;
+      case 9: // Uint64
+        final view = typedValues.cast<ffi.Uint64>().asTypedList(nValues);
+        return copy ? Uint64List.fromList(view) : view;
+      case 10: // Float32
+        final view = typedValues.cast<ffi.Float>().asTypedList(nValues);
+        return copy ? Float32List.fromList(view) : view;
+      case 11: // Float64
+        final view = typedValues.cast<ffi.Double>().asTypedList(nValues);
+        return copy ? Float64List.fromList(view) : view;
```

Comment:
nit: we can extract the common pattern to make it a bit cleaner

Reply:
Actually I considered it and figured probably not: type casts are specific and let us access dart extensions to typed lists based on ffi type. Moreover, even though this looks redundant, it's clear and statically well typed.

However if you have a better solution that preserves the same level of type checking I sure am taking the change!

Reply:
e.g.

```
      case 9: // Uint64
        return f<ffi.Uint64, Uint64List>(typedvalues, nvalues, copy, Uint64List.fromList);
      case 10: // Float32
        return f<ffi.float32, Float32list>(typedvalues, nvalues, copy, float32List.fromList);

S f<T, S>(typedValues, S Function() copyer, bool copy) {
          final view = typedValues.cast<T>().asTypedList(nValues);
        return copy ? copyer(view) : view;
}
```

Reply:
Won't be enough: `asTypedList` extension only exists for the specific type of each case. I doubt dart's type system is strong enough to let us abstract the union of these types + the corresponding copyer correctly. Especially that `fromList` is static and that dart won't let us interface static functions.

I'd still buy working implementation though :smile:

Reply:
Anyway, as long as you add one test for each case in `main.dart` and verify the types and content, then we are safe